File: mobile/src/main/java/com/alexstyl/specialdates/events/peopleevents/PeopleNamedaysCalculator.java
package com.alexstyl.specialdates.events.peopleevents;

import com.alexstyl.specialdates.DisplayName;
import com.alexstyl.specialdates.Optional;
import com.alexstyl.specialdates.contact.Contact;
import com.alexstyl.specialdates.contact.AndroidContactsProvider;
import com.alexstyl.specialdates.date.ContactEvent;
import com.alexstyl.specialdates.date.Date;
import com.alexstyl.specialdates.events.namedays.NameCelebrations;
import com.alexstyl.specialdates.events.namedays.NamedayLocale;
import com.alexstyl.specialdates.events.namedays.NamedayPreferences;
import com.alexstyl.specialdates.events.namedays.calendar.NamedayCalendar;
import com.alexstyl.specialdates.events.namedays.calendar.resource.NamedayCalendarProvider;
import com.alexstyl.specialdates.date.TimePeriod;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;

public final class PeopleNamedaysCalculator {

    private static final Optional<Long> NO_DEVICE_EVENT_ID = Optional.absent();
    private final NamedayPreferences namedayPreferences;
    private final NamedayCalendarProvider namedayCalendarProvider;
    private final AndroidContactsProvider contactsProvider;

    public PeopleNamedaysCalculator(NamedayPreferences namedayPreferences,
                                    NamedayCalendarProvider namedayCalendarProvider,
                                    AndroidContactsProvider contactsProvider) {
        this.namedayPreferences = namedayPreferences;
        this.namedayCalendarProvider = namedayCalendarProvider;
        this.contactsProvider = contactsProvider;
    }

    public List<ContactEvent> loadDeviceStaticNamedays() {
        List<ContactEvent> namedayEvents = new ArrayList<>();
        List<Contact> contacts = contactsProvider.fetchAllDeviceContacts();
        for (Contact contact : contacts) {
            DisplayName displayName = contact.getDisplayName();
            HashSet<Date> namedays = new HashSet<>();
            for (String firstName : displayName.getFirstNames()) {
                NameCelebrations nameDays = getNamedaysOf(firstName);
                if (nameDays.containsNoDate()) {
                    continue;
                }
                int namedaysCount = nameDays.size();
                for (int i = 0; i < namedaysCount; i++) {
                    Date date = nameDays.getDate(i);
                    if (namedays.contains(date)) {
                        continue;
                    }
                    ContactEvent event = new ContactEvent(NO_DEVICE_EVENT_ID, StandardEventType.NAMEDAY, date, contact);
                    namedayEvents.add(event);
                    namedays.add(date);
                }
            }
        }

        return namedayEvents;
    }

    public List<ContactEvent> loadSpecialNamedaysBetween(TimePeriod timeDuration) {
        List<ContactEvent> namedayEvents = new ArrayList<>();
        for (Contact contact : contactsProvider.fetchAllDeviceContacts()) {
            for (String firstName : contact.getDisplayName().getFirstNames()) {
                NameCelebrations nameDays = getSpecialNamedaysOf(firstName);
                if (nameDays.containsNoDate()) {
                    continue;
                }

                int namedaysCount = nameDays.size();
                for (int i = 0; i < namedaysCount; i++) {
                    Date date = nameDays.getDate(i);
                    if (timeDuration.containsDate(date)) {
                        ContactEvent nameday = new ContactEvent(NO_DEVICE_EVENT_ID, StandardEventType.NAMEDAY, date, contact);
                        namedayEvents.add(nameday);
                    }
                }
            }
        }
        return namedayEvents;
    }

    private NameCelebrations getNamedaysOf(String given) {
        NamedayCalendar namedayCalendar = getNamedayCalendar();
        return namedayCalendar.getNormalNamedaysFor(given);
    }

    private NameCelebrations getSpecialNamedaysOf(String firstName) {
        NamedayCalendar namedayCalendar = getNamedayCalendar();
        return namedayCalendar.getSpecialNamedaysFor(firstName);
    }

    public NamedayCalendar getNamedayCalendar() {
        NamedayLocale locale = namedayPreferences.getSelectedLanguage();
        return namedayCalendarProvider.loadNamedayCalendarForLocale(locale, Date.CURRENT_YEAR);
    }
}

File: memento/src/main/java/com/alexstyl/specialdates/events/peopleevents/PeopleNamedaysCalculator.kt
package com.alexstyl.specialdates.events.peopleevents

import com.alexstyl.specialdates.Optional
import com.alexstyl.specialdates.contact.Contact
import com.alexstyl.specialdates.contact.ContactsProvider
import com.alexstyl.specialdates.date.ContactEvent
import com.alexstyl.specialdates.date.Date
import com.alexstyl.specialdates.date.DateComparator
import com.alexstyl.specialdates.date.TimePeriod
import com.alexstyl.specialdates.events.namedays.NameCelebrations
import com.alexstyl.specialdates.events.namedays.NamedayUserSettings
import com.alexstyl.specialdates.events.namedays.calendar.NamedayCalendar
import com.alexstyl.specialdates.events.namedays.calendar.resource.NamedayCalendarProvider

open class PeopleNamedaysCalculator(
        private val namedayPreferences: NamedayUserSettings,
        private val namedayCalendarProvider: NamedayCalendarProvider,
        private val contactsProvider: ContactsProvider) : PeopleEventsProvider {

    private val namedayCalendar: NamedayCalendar
        get() {
            val locale = namedayPreferences.selectedLanguage
            return namedayCalendarProvider.loadNamedayCalendarForLocale(locale, Date.CURRENT_YEAR)
        }

    override fun fetchEventsOn(date: Date): ContactEventsOnADate {
        val contactEvents = fetchEventsBetween(TimePeriod.between(date, date))
        return ContactEventsOnADate.createFrom(date, contactEvents)
    }

    override fun fetchEventsBetween(timePeriod: TimePeriod): List<ContactEvent> {
        val namedayEvents = ArrayList<ContactEvent>()
        for (contact in contactsProvider.allContacts) {
            for (firstName in contact.displayName.firstNames) {
                val nameDays = getSpecialNamedaysOf(firstName)
                if (nameDays.containsNoDate()) {
                    continue
                }

                val namedaysCount = nameDays.size()
                for (i in 0 until namedaysCount) {
                    val date = nameDays.getDate(i)
                    if (timePeriod.containsDate(date)) {
                        val nameday = ContactEvent(Optional(contact.contactID), StandardEventType.NAMEDAY, date, contact)
                        namedayEvents.add(nameday)
                    }
                }
            }
        }
        return namedayEvents.toList()
    }

    override fun fetchEventsFor(contact: Contact): List<ContactEvent> {
        val namedays = ArrayList<ContactEvent>()
        for (name in contact.displayName.allNames) {
            val (_, specialDates) = getSpecialNamedaysOf(name)
            for (i in 0 until specialDates.size()) {
                val date = specialDates.getDate(i)
                val deviceEventId = Optional(contact.contactID)
                val contactEvent = ContactEvent(deviceEventId, StandardEventType.NAMEDAY, date, contact)
                namedays.add(contactEvent)
            }
        }
        return namedays
    }

    @Throws(NoEventsFoundException::class)
    override fun findClosestEventDateOnOrAfter(date: Date): Date {
        val timePeriod = TimePeriod.between(date, Date.endOfYear(date.getYear()))
        val contactEvents = ArrayList(fetchEventsBetween(timePeriod))
        contactEvents.sortWith(Comparator { (_, _, date1), (_, _, date2) -> DATE_COMPARATOR.compare(date1, date2) })

        for ((_, _, contactEventDate) in contactEvents) {
            if (DATE_COMPARATOR.compare(contactEventDate, date) >= 0) {
                return contactEventDate
            }
        }
        throw NoEventsFoundException("No nameday events found on or after $date")
    }

    fun loadDeviceStaticNamedays(): List<ContactEvent> {
        val namedayEvents = ArrayList<ContactEvent>()
        val contacts = contactsProvider.allContacts
        for (contact in contacts) {
            val displayName = contact.displayName
            val namedays = HashSet<Date>()
            for (firstName in displayName.firstNames) {
                val nameDays = getNamedaysOf(firstName)
                if (nameDays.containsNoDate()) {
                    continue
                }
                val namedaysCount = nameDays.size()
                for (i in 0 until namedaysCount) {
                    val date = nameDays.getDate(i)
                    if (namedays.contains(date)) {
                        continue
                    }
                    val event = ContactEvent(Optional(contact.contactID), StandardEventType.NAMEDAY, date, contact)
                    namedayEvents.add(event)
                    namedays.add(date)
                }
            }
        }

        return namedayEvents
    }

    private fun getNamedaysOf(given: String): NameCelebrations {
        val namedayCalendar = namedayCalendar
        return namedayCalendar.getNormalNamedaysFor(given)
    }

    private fun getSpecialNamedaysOf(firstName: String): NameCelebrations {
        val namedayCalendar = namedayCalendar
        return namedayCalendar.getSpecialNamedaysFor(firstName)
    }

    companion object {

        private val DATE_COMPARATOR = DateComparator.INSTANCE
    }

}
